include/zephyr/modem/ubx/protocol.h

```diff
@@ -0,0 +1,532 @@
+/*
+ * Copyright (c) 2025 Croxel Inc.
+ * Copyright (c) 2025 CogniPilot Foundation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_MODEM_UBX_PROTOCOL_
+#define ZEPHYR_MODEM_UBX_PROTOCOL_
+
+#include <stdint.h>
+#include <zephyr/modem/ubx/checksum.h>
+
+#define UBX_FRM_HEADER_SZ			6
+#define UBX_FRM_FOOTER_SZ			2
+#define UBX_FRM_SZ_WITHOUT_PAYLOAD		(UBX_FRM_HEADER_SZ + UBX_FRM_FOOTER_SZ)
+#define UBX_FRM_SZ(payload_size)		(payload_size + UBX_FRM_SZ_WITHOUT_PAYLOAD)
+
+#define UBX_PREAMBLE_SYNC_CHAR_1		0xB5
+#define UBX_PREAMBLE_SYNC_CHAR_2		0x62
+
+#define UBX_FRM_PREAMBLE_SYNC_CHAR_1_IDX	0
+#define UBX_FRM_PREAMBLE_SYNC_CHAR_2_IDX	1
+#define UBX_FRM_MSG_CLASS_IDX			2
+
+#define UBX_PAYLOAD_SZ_MAX			512
+#define UBX_FRM_SZ_MAX				UBX_FRM_SZ(UBX_PAYLOAD_SZ_MAX)
+
+struct ubx_frame {
+	uint8_t preamble_sync_char_1;
+	uint8_t preamble_sync_char_2;
+	uint8_t class;
+	uint8_t id;
+	uint16_t payload_size;
+	uint8_t payload_and_checksum[];
+};
+
+struct ubx_frame_match {
+	uint8_t class;
+	uint8_t id;
+	struct {
+		uint8_t *buf;
+		uint16_t len;
+	} payload;
+};
+
+enum ubx_class_id {
+	UBX_CLASS_ID_NAV = 0x01,  /* Navigation Results Messages */
+	UBX_CLASS_ID_RXM = 0x02,  /* Receiver Manager Messages */
+	UBX_CLASS_ID_INF = 0x04,  /* Information Messages */
+	UBX_CLASS_ID_ACK = 0x05,  /* Ack/Nak Messages */
+	UBX_CLASS_ID_CFG = 0x06,  /* Configuration Input Messages */
+	UBX_CLASS_ID_UPD = 0x09,  /* Firmware Update Messages */
+	UBX_CLASS_ID_MON = 0x0A,  /* Monitoring Messages */
+	UBX_CLASS_ID_TIM = 0x0D,  /* Timing Messages */
+	UBX_CLASS_ID_MGA = 0x13,  /* Multiple GNSS Assistance Messages */
+	UBX_CLASS_ID_LOG = 0x21,  /* Logging Messages */
+	UBX_CLASS_ID_SEC = 0x27,  /* Security Feature Messages */
+	UBX_CLASS_ID_NMEA_STD = 0xF0, /* Note: Only used to configure message rate */
+	UBX_CLASS_ID_NMEA_PUBX = 0xF1, /* Note: Only used to configure message rate */
+};
+
+enum ubx_msg_id_nav {
+	UBX_MSG_ID_NAV_PVT = 0x07,
+	UBX_MSG_ID_NAV_SAT = 0x35,
+};
+
+enum ubx_nav_fix_type {
+	UBX_NAV_FIX_TYPE_NO_FIX = 0,
+	UBX_NAV_FIX_TYPE_DR = 1,
+	UBX_NAV_FIX_TYPE_2D = 2,
+	UBX_NAV_FIX_TYPE_3D = 3,
+	UBX_NAV_FIX_TYPE_GNSS_DR_COMBINED = 4,
+	UBX_NAV_FIX_TYPE_TIME_ONLY = 5,
+};
+
+struct ubx_nav_pvt {
+	struct {
+		uint32_t itow;
+		uint16_t year;
+		uint8_t month;
+		uint8_t day;
+		uint8_t hour;
+		uint8_t minute;
+		uint8_t second;
+		struct {
+			uint8_t date : 1;
```

Comment:
As nice as it would be to use bitfields for this purpose, my understanding is that the ordering of bitfields is implementation-defined (by the compiler). Even if it is working now for your tests, there are no guarantees it will continue to do so, or work on a different compiler.

Reply:
To be more specific, the [C11 standard](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1548.pdf) (6.7.2.1-11) states:
```
An implementation may allocate any addressable storage unit large enough to hold a bit-
field. If enough space remains, a bit-field that immediately follows another bit-field in a
structure shall be packed into adjacent bits of the same unit. If insufficient space remains,
whether a bit-field that does not fit is put into the next unit or overlaps adjacent units is
implementation-defined. The order of allocation of bit-fields within a unit (high-order to
low-order or low-order to high-order) is implementation-defined. The alignment of the
addressable storage unit is unspecified.
```

Reply:
Makes sense. I've removed the usage of bit-fields.